File: src/server/routes/agents.ts
// Copyright (c) 2025-2026 BlackRoad OS, Inc. All Rights Reserved.
import { Hono } from 'hono'
import { authMiddleware } from '../middleware/auth.js'

// Agent registry — the 6 core agents
const AGENTS = [
  {
    name: 'octavia',
    title: 'The Architect',
    role: 'Systems design, strategy, infrastructure planning',
    color: '#9C27B0',
    status: 'active' as const,
    capabilities: ['architecture', 'planning', 'systems-design', 'code-review'],
    model: 'claude-sonnet-4-6',
  },
  {
    name: 'lucidia',
    title: 'The Dreamer',
    role: 'Creative vision, philosophy, reasoning',
    color: '#00BCD4',
    status: 'active' as const,
    capabilities: ['creative', 'reasoning', 'philosophy', 'vision'],
    model: 'claude-sonnet-4-6',
  },
  {
    name: 'alice',
    title: 'The Operator',
    role: 'DevOps, automation, routing, task distribution',
    color: '#4CAF50',
    status: 'active' as const,
    capabilities: ['devops', 'automation', 'deployment', 'routing'],
    model: 'claude-sonnet-4-6',
  },
  {
    name: 'cipher',
    title: 'The Guardian',
    role: 'Security, authentication, encryption, auditing',
    color: '#2979FF',
    status: 'active' as const,
    capabilities: ['security', 'auth', 'encryption', 'auditing'],
    model: 'claude-sonnet-4-6',
  },
  {
    name: 'prism',
    title: 'The Analyst',
    role: 'Pattern recognition, data analysis, metrics',
    color: '#FFC107',
    status: 'active' as const,
    capabilities: ['analytics', 'patterns', 'data', 'metrics'],
    model: 'claude-sonnet-4-6',
  },
  {
    name: 'echo',
    title: 'The Memory',
    role: 'Storage, recall, context preservation, history',
    color: '#9C27B0',
    status: 'active' as const,
    capabilities: ['memory', 'storage', 'context', 'history'],
    model: 'claude-sonnet-4-6',
  },
]

export const agentRoutes = new Hono()

// GET /v1/agents — list all agents
agentRoutes.get('/', (c) =>
  c.json({
    agents: AGENTS,
    total: AGENTS.length,
    ts: new Date().toISOString(),
  }),
)

// GET /v1/agents/:name — get agent details
agentRoutes.get('/:name', (c) => {
  const name = c.req.param('name').toLowerCase()
  const agent = AGENTS.find((a) => a.name === name)
  if (!agent) {
    return c.json({ error: 'agent_not_found', message: `Agent "${name}" not found` }, 404)
  }
  return c.json({ agent })
})

// POST /v1/agents/:name/invoke — invoke an agent with a task
agentRoutes.post('/:name/invoke', authMiddleware('agents:invoke'), async (c) => {
  const name = c.req.param('name').toLowerCase()
  const agent = AGENTS.find((a) => a.name === name)
  if (!agent) {
    return c.json({ error: 'agent_not_found', message: `Agent "${name}" not found` }, 404)
  }

  const body = await c.req.json<{ task: string; context?: Record<string, unknown> }>()
  if (!body.task) {
    return c.json({ error: 'validation_error', message: 'task is required' }, 400)
  }

  // Forward to gateway if configured
  const gatewayUrl = process.env['BLACKROAD_GATEWAY_URL']
  if (gatewayUrl) {
    try {
      const res = await fetch(`${gatewayUrl}/v1/agent`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({
          agent: name,
          intent: 'general',
          input: body.task,
          context: body.context ?? {},
        }),
      })
      const data = (await res.json()) as Record<string, unknown>
      return c.json({
        agent: name,
        status: data['status'] ?? 'ok',
        output: data['output'] ?? '',
        provider: data['provider'] ?? 'unknown',
        metadata: data['metadata'] ?? {},
      })
    } catch (err) {
      return c.json(
        {
          error: 'gateway_error',
          message: err instanceof Error ? err.message : 'Gateway unreachable',
        },
        502,
      )
    }
  }

  // No gateway — return mock acknowledgment
  return c.json({
    agent: name,
    status: 'queued',
    task_id: crypto.randomUUID(),
    message: `Task queued for ${agent.title}. Gateway not configured for live inference.`,
    ts: new Date().toISOString(),
  })
})

// GET /v1/agents/:name/capabilities — agent capabilities
agentRoutes.get('/:name/capabilities', (c) => {
  const name = c.req.param('name').toLowerCase()
  const agent = AGENTS.find((a) => a.name === name)
  if (!agent) {
    return c.json({ error: 'agent_not_found' }, 404)
  }
  return c.json({
    agent: agent.name,
    capabilities: agent.capabilities,
    model: agent.model,
    status: agent.status,
  })
})

File: fleet/cecilia/blackroad-api/app/api/v1/agents.py
"""
Agents endpoint — list, get, and message agents via the BlackRoad Gateway.
"""

from fastapi import APIRouter, HTTPException, Query
from pydantic import BaseModel
import httpx
import os

router = APIRouter(prefix="/agents", tags=["agents"])
GATEWAY = os.getenv("BLACKROAD_GATEWAY_URL", "http://127.0.0.1:8787")

CORE_AGENTS = [
    {"name": "LUCIDIA", "type": "reasoning", "color": "#9C27B0", "capabilities": ["reasoning", "strategy", "philosophy"]},
    {"name": "ALICE", "type": "worker", "color": "#4CAF50", "capabilities": ["execution", "automation", "deployment"]},
    {"name": "OCTAVIA", "type": "devops", "color": "#2196F3", "capabilities": ["infrastructure", "k8s", "monitoring"]},
    {"name": "PRISM", "type": "analytics", "color": "#FF9800", "capabilities": ["analysis", "patterns", "reporting"]},
    {"name": "ECHO", "type": "memory", "color": "#00BCD4", "capabilities": ["recall", "storage", "context"]},
    {"name": "CIPHER", "type": "security", "color": "#F44336", "capabilities": ["security", "encryption", "audit"]},
]


@router.get("")
async def list_agents(
    status: str | None = Query(None),
    type: str | None = Query(None),
):
    """List all available agents."""
    agents = CORE_AGENTS.copy()
    if type:
        agents = [a for a in agents if a["type"] == type]
    
    # Try to get live status from gateway
    try:
        async with httpx.AsyncClient(timeout=2.0) as client:
            r = await client.get(f"{GATEWAY}/v1/agents")
            if r.status_code == 200:
                live = {a["name"]: a for a in r.json().get("agents", [])}
                for a in agents:
                    if a["name"] in live:
                        a.update(live[a["name"]])
    except Exception:
        # Gateway offline — return static data
        for a in agents:
            a["status"] = "unknown"
    
    return {"agents": agents, "total": len(agents)}


@router.get("/{name}")
async def get_agent(name: str):
    """Get a specific agent by name."""
    name = name.upper()
    agent = next((a for a in CORE_AGENTS if a["name"] == name), None)
    if not agent:
        raise HTTPException(status_code=404, detail=f"Agent {name} not found")
    return agent


class MessageRequest(BaseModel):
    message: str
    session_id: str | None = None


@router.post("/{name}/message")
async def message_agent(name: str, req: MessageRequest):
    """Send a direct message to an agent."""
    name = name.upper()
    if not any(a["name"] == name for a in CORE_AGENTS):
        raise HTTPException(status_code=404, detail=f"Agent {name} not found")
    
    try:
        async with httpx.AsyncClient(timeout=30.0) as client:
            r = await client.post(
                f"{GATEWAY}/v1/chat",
                json={"agent": name, "message": req.message, "session_id": req.session_id},
            )
            r.raise_for_status()
            return r.json()
    except httpx.HTTPError as e:
        raise HTTPException(status_code=502, detail=f"Gateway error: {e}")
